### ISSUE TITLE
Allow process method of block extension to return any abstract block

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#4234]
If the process method of a block extension returns an abstract block with the context `:compound`, but is not an instance of `Block`, Asciidoctor crashes. This happens because it tries to access the `lines` method, which is only available on `Block`. This prevents the extension from being able to use the return value to insert a list, section, or table.

```ruby
block do
  named :ul
  on_context :paragraph
  process do |parent, reader|
    reader.read_lines.each_with_object create_list parent, :ulist do |line, list|
      list << (create_list_item list, line)
    end
  end
end
```

The parser should only attempt to access the `lines` method if the return value is an instance of `Block`.

Furthermore, the parser should treat the return value as `nil` if the value matches the parent argument (to avoid recursing into itself). The latter change should be made for a custom block macro as well. This change allows an extension to safely use the `parse_content` method without having to worry about the return value.

```rb
block do
  named :unwrap
  on_context :open
  process do |parent, reader|
    parse_content parent, reader.read_lines
  end
end
```

Note that the parser will still assume that the return value is an abstract block or nil. It will not attempt to check for other, unexpected values not defined by the contract.

**Comment by mojavelinux** [comment]
Backport this change to v2.0.x.

**Comment by mojavelinux** [comment]
Resolved by de1ba697bdbc8da1a6598269f4b0093c83887f5b
